File: backend/src/main/kotlin/com/trackwatch/backend/service/TrackService.kt
package com.trackwatch.backend.service

import com.trackwatch.backend.clients.spotify.SpotifyArtistApiClient
import com.trackwatch.backend.model.Artist
import com.trackwatch.backend.model.Track
import com.trackwatch.backend.model.User
import com.trackwatch.backend.utils.values.Constants
import org.springframework.stereotype.Service
import java.util.*
import kotlin.math.abs

@Service
class TrackService(private val spotifyArtistApiClient: SpotifyArtistApiClient) {

    fun searchArtistTracks(
        artist: Artist,
        accessToken: String,
        pagesToFetch: Int,
        daysLimit: Int = Constants.FILTER_DAYS_LIMIT
    ): List<Track> {
        val newTracks = mutableListOf<Track>()

        for (page in 0 until pagesToFetch) {
            val pageNewTracks =
                spotifyArtistApiClient.searchArtistTracksWithRetries(artist, accessToken, daysLimit, page)
            newTracks.addAll(pageNewTracks)
        }

        return newTracks
    }

    fun filterTrack(
        track: Track,
        user: User,
        artist: Artist,
        tracksToAdd: MutableSet<Track>,
        daysLimit: Int = Constants.FILTER_DAYS_LIMIT,
        shouldCheckCorrectArtist: Boolean = true,
        shouldCheckTrackInTimeRange: Boolean = true,
        shouldCheckCompilationAlbum: Boolean = true,
        shouldCheckSongBlockedByUserSettings: Boolean = true,
        shouldCheckTrackRecentlyAdded: Boolean = false
    ): Track? {
        val calendar = Calendar.getInstance(TimeZone.getTimeZone(Constants.SERVER_TIMEZONE))
        val today = calendar.time

        calendar.add(Calendar.DAY_OF_YEAR, -daysLimit)
        val startDate = calendar.time

        val isCorrectArtist = !shouldCheckCorrectArtist || isCorrectArtist(track, artist)
        val isTrackInTimeRange = !shouldCheckTrackInTimeRange || isTrackInTimeRange(track, startDate, today)
        val isCompilationAlbum = shouldCheckCompilationAlbum && isCompilationAlbum(track)
        val isSongBlockedByUserSettings =
            shouldCheckSongBlockedByUserSettings && isSongBlockedByUserSettings(track, user)
        val isTrackRecentlyAdded = shouldCheckTrackRecentlyAdded && isTrackRecentlyAdded(user, track)

        if (
            isCorrectArtist &&
            isTrackInTimeRange &&
            !isCompilationAlbum &&
            !isSongBlockedByUserSettings &&
            !isTrackRecentlyAdded
        ) {
            val selectedTrack = selectTrack(track, tracksToAdd)
            if (isSameTrackInList(selectedTrack, tracksToAdd)) {
                return null
            }
            tracksToAdd.add(selectedTrack)
            return selectedTrack
        }

        return null
    }

    fun sortTracks(tracks: Set<Track>): Set<Track> {
        val sortedList = tracks.sortedWith(
            compareBy(
                { it.releaseDate },
                { it.albumName },
                { it.discNumber },
                { it.albumOrder })
        )
        return sortedList.toSet()
    }

    fun removeDuplicateTracks(tracks: List<Track>): List<Track> {
        val uniqueTracks = mutableListOf<Track>()
        val processedTracks = mutableMapOf<String, Track>()

        for (track in tracks) {
            val baseSignature = generateTrackSignatureWithoutDuration(track)

            if (processedTracks.containsKey(baseSignature)) {
                val existingTrack = processedTracks[baseSignature]!!
                val durationDifference = abs(track.durationMs - existingTrack.durationMs)

                // If duration differs by more than 1000ms (1 second), consider it a different track
                if (durationDifference > 1000) {
                    val uniqueSignature = "$baseSignature|${track.durationMs}"
                    processedTracks[uniqueSignature] = track
                    uniqueTracks.add(track)
                }
            } else {
                processedTracks[baseSignature] = track
                uniqueTracks.add(track)
            }
        }

        return uniqueTracks
    }

    private fun generateTrackSignatureWithoutDuration(track: Track): String {
        val normalizedName = track.name.lowercase()
        val artistsSignature = track.artists
            .map { it.name.lowercase() }
            .sorted()
            .joinToString(",")

        return "$normalizedName|$artistsSignature"
    }

    fun isCorrectArtist(track: Track, artist: Artist): Boolean {
        return track.artists.any { it.id == artist.id }
    }

    fun isTrackInTimeRange(track: Track, startDate: Date, endDate: Date): Boolean {
        return track.releaseDate in startDate..endDate
    }

    fun isCompilationAlbum(track: Track): Boolean {
        return track.albumType == "compilation"
    }

    fun isSongBlockedByUserSettings(track: Track, user: User): Boolean {
        return user.settings.blockedExplicitContent && track.isExplicit
    }

    private fun selectTrack(track: Track, tracksToAdd: Set<Track>): Track {
        val equalTrack = tracksToAdd.find { it.isEqualTo(track) }
        if (equalTrack == null) {
            return track
        }

        var selectedTrack = track
        var nonSelectedTrack = equalTrack

        // Prefer album tracks over single tracks
        if (selectedTrack.albumType == "single" && equalTrack.albumType == "album") {
            selectedTrack = equalTrack
            nonSelectedTrack = track
        }

        // Prefer explicit tracks over non-explicit tracks
        if (!selectedTrack.isExplicit && nonSelectedTrack.isExplicit) {
            selectedTrack = nonSelectedTrack
        }

        return selectedTrack
    }

    private fun isSameTrackInList(track: Track, tracks: Set<Track>): Boolean {
        return tracks.any { it.isEqualStrict(track) }
    }

    private fun isTrackRecentlyAdded(user: User, track: Track): Boolean {
        val recentlyAddedTracks = user.recentlyAddedTracks
        return recentlyAddedTracks.any { it.isEqualToTrack(track) }
    }

}

File: backend_py/trackwatch/app/services/track_service.py
import datetime
from app.constants import System
from app.clients.spotify.spotify_artist_api_client import search_artist_tracks_with_retries

def search_artist_tracks(
  artist,
  access_token: str,
  pages_to_fetch: int,
  days_limit: int = System.FILTER_DAYS_LIMIT
) -> list:
  new_tracks = []

  for page in range(pages_to_fetch):
    page_new_tracks = search_artist_tracks_with_retries(
      artist,
      access_token,
      days_limit,
      page
    )
    new_tracks.extend(page_new_tracks)

  return new_tracks

def filter_track(
  track,
  user,
  artist,
  tracks_to_add: set,
  days_limit: int = System.FILTER_DAYS_LIMIT,
  should_check_correct_artist: bool = True,
  should_check_track_in_time_range: bool = True,
  should_check_compilation_album: bool = True,
  should_check_song_blocked_by_user_settings: bool = True,
  should_check_track_recently_added: bool = False
):
  tz = datetime.timezone(datetime.timedelta(hours=int(System.SERVER_TIMEZONE)))
  today = datetime.datetime.now(tz)
  start_date = today - datetime.timedelta(days=days_limit)

  is_correct_artist_result = (not should_check_correct_artist) or is_correct_artist(track, artist)
  is_track_in_time_range_result = (not should_check_track_in_time_range) or is_track_in_time_range(track, start_date, today)
  is_compilation_album_result = should_check_compilation_album and is_compilation_album(track)
  is_song_blocked_result = should_check_song_blocked_by_user_settings and is_song_blocked_by_user_settings(track, user)
  is_track_recently_added_result = should_check_track_recently_added and is_track_recently_added(user, track)

  if (
    is_correct_artist_result and
    is_track_in_time_range_result and
    not is_compilation_album_result and
    not is_song_blocked_result and
    not is_track_recently_added_result
  ):
    selected_track = select_track(track, tracks_to_add)
    if is_same_track_in_list(selected_track, tracks_to_add): return None
    tracks_to_add.add(selected_track)
    return selected_track

  return None

def sort_tracks(tracks: set) -> set:
  sorted_list = sorted(
    tracks,
    key=lambda t: (
      t.release_date,
      t.album_name,
      t.disc_number,
      t.album_order
    )
  )
  return set(sorted_list)

def remove_duplicate_tracks(tracks: list) -> list:
  unique_tracks = []
  processed_tracks = {}

  for track in tracks:
    base_signature = generate_track_signature_without_duration(track)

    if base_signature in processed_tracks:
      existing_track = processed_tracks[base_signature]
      duration_difference = abs(track.duration_ms - existing_track.duration_ms)

      if duration_difference > 1000:
        unique_signature = f"{base_signature}|{track.duration_ms}"
        processed_tracks[unique_signature] = track
        unique_tracks.append(track)

    else:
      processed_tracks[base_signature] = track
      unique_tracks.append(track)

  return unique_tracks

def generate_track_signature_without_duration(track) -> str:
  normalized_name = track.name.lower()
  artists_signature = ",".join(sorted([a.name.lower() for a in track.artists]))
  return f"{normalized_name}|{artists_signature}"

def is_correct_artist(track, artist) -> bool:
  return any(a.id == artist.id for a in track.artists)

def is_track_in_time_range(track, start_date, end_date) -> bool:
  return start_date <= track.release_date <= end_date

def is_compilation_album(track) -> bool:
  return getattr(track, "album_type", "") == "compilation"

def is_song_blocked_by_user_settings(track, user) -> bool:
  blocked = getattr(user, "settings", None)
  if blocked: return getattr(blocked, "blocked_explicit_content", False) and track.is_explicit

  return user.settings.get("blocked_explicit_content", False) and track.is_explicit

def select_track(track, tracks_to_add: set):
  equal_track = next((t for t in tracks_to_add if t.is_equal_to(track)), None)
  if equal_track is None: return track

  selected_track = track
  non_selected_track = equal_track

  if selected_track.album_type == "single" and equal_track.album_type == "album":
    selected_track = equal_track
    non_selected_track = track

  if not selected_track.is_explicit and non_selected_track.is_explicit:
    selected_track = non_selected_track

  return selected_track

def is_same_track_in_list(track, tracks: set) -> bool:
  return any(t.is_equal_strict(track) for t in tracks)

def is_track_recently_added(user, track) -> bool:
  recently_added = getattr(user, "recently_added_tracks", [])
  return any(t.is_equal_to(track) for t in recently_added)
